drv/stm32h7-spi-rot-server/src/main.rs

```diff
@@ -0,0 +1,274 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use idol_runtime::{ClientError, Leased, RequestError, R, W};
+// use idol_runtime::{NotificationHandler, ClientError, Leased, RequestError, R, W};
+use drv_spi_api::{CsState, Spi, SpiError};
+use drv_spi_msg::*;
+use drv_stm32xx_sys_api as sys_api;
+use userlib::*;
+
+task_slot!(SPI, spi_driver);
+task_slot!(SYS, sys);
+
+use ringbuf::*;
+
+const SPI_TO_ROT_DEVICE: u8 = 0;
+
+// On Gemini, the STM32H753 is in a LQFP176 package with ROT_IRQ on pin2/PE3
+const ROT_IRQ: sys_api::PinSet = sys_api::PinSet {
+    port: sys_api::Port::E,
+    pin_mask: 1 << 3,
+};
+
+// const PACKET_MASK: u32 = 1 << 0;
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+enum Trace {
+    Init,
+    BadMessageLength(usize),
+    BadProtocol(u8),
+    RspLen(usize),
+    SpiError(SpiError),
+    Data(u8, u8, u8, u8),
+    GpioPort(u16, u16),
+    Line,
+    ReturnOk(MsgType, u32),
+    RotIrqTimeout,
+    RotIrqAsserted,
+    SendRecv(MsgType, usize, usize),
+}
+ringbuf!(Trace, 32, Trace::Init);
+
+#[export_name = "main"]
+fn main() -> ! {
+    let spi = Spi::from(SPI.get_task_id()).device(SPI_TO_ROT_DEVICE);
+    let sys = sys_api::Sys::from(SYS.get_task_id());
+
+    sys.gpio_configure_input(ROT_IRQ, sys_api::Pull::None)
+        .unwrap_lite();
+
+    let mut buffer = [0; idl::INCOMING_SIZE];
+    // The larger of the two buffers.
+    let message = &mut [0u8; drv_spi_msg::SPI_RSP_BUF_SIZE];
+    let mut server = ServerImpl {
+        sys,
+        spi,
+        message: *message,
+    };
+
+    loop {
+        idol_runtime::dispatch(&mut buffer, &mut server);
+    }
+}
+
+fn do_send_recv(server: &mut ServerImpl) -> Result<usize, MsgError> {
```

Comment:
A docstring here would be nice, explaining what this function is supposed to do.